Create azure App Service Plan resource

lstolyarov opened the pull request.

lstolyarov commented:
@tombuildsstuff I have migrated the code in pull requests https://github.com/hashicorp/terraform/pull/12001 and https://github.com/hashicorp/terraform/pull/13634 to this repository.

tombuildsstuff commented on azurerm/resource_arm_app_service_plan.go:
```diff
@@ -0,0 +1,133 @@
+	client := meta.(*ArmClient)
+	AppServicePlanClient := client.appServicePlansClient
+
+	log.Printf("[INFO] preparing arguments for Azure App Service Plan creation.")
```
can we update this to `AzureRM` for consistency?

tombuildsstuff commented on azurerm/resource_arm_app_service_plan.go:
```diff
@@ -0,0 +1,133 @@
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"location": {
```
can we switch this to use `locationSchema()`?

tombuildsstuff commented on azurerm/resource_arm_app_service_plan.go:
```diff
@@ -0,0 +1,133 @@
+
+func resourceArmAppServicePlanCreateUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient)
+	AppServicePlanClient := client.appServicePlansClient
```
_minor_: we'd generally combine these into one variable `client`

tombuildsstuff commented on azurerm/resource_arm_app_service_plan.go:
```diff
@@ -0,0 +1,133 @@
+		Create: resourceArmAppServicePlanCreateUpdate,
+		Read:   resourceArmAppServicePlanRead,
+		Update: resourceArmAppServicePlanCreateUpdate,
+		Delete: resourceArmAppServicePlanDelete,
```
given there's a test for it, can we add import support here:

```
Importer: &schema.ResourceImporter{
    State: schema.ImportStatePassthrough,
},
```

tombuildsstuff commented on azurerm/resource_arm_app_service_plan.go:
```diff
@@ -0,0 +1,133 @@
+			},
+			"maximum_number_of_workers": {
+				Type:     schema.TypeString,
+				Optional: true,
```
should this also be `Computed`?

tombuildsstuff commented on azurerm/resource_arm_app_service_plan.go:
```diff
@@ -0,0 +1,133 @@
+				Required: true,
+			},
+			"name": {
+				Type:     schema.TypeString,
```
I believe this should be ForceNew?

tombuildsstuff commented on azurerm/resource_arm_app_service_plan.go:
```diff
@@ -0,0 +1,133 @@
+		Schema: map[string]*schema.Schema{
+			"resource_group_name": {
+				Type:     schema.TypeString,
+				Required: true,
```
I believe this should be ForceNew?

tombuildsstuff commented on azurerm/resource_arm_app_service_plan.go:
```diff
@@ -0,0 +1,133 @@
+
+	properties := web.AppServicePlanProperties{}
+	if v, ok := d.GetOk("maximum_number_of_workers"); ok {
+		maximumNumberOfWorkers := v.(int32)
```
can this be cast directly, or does it need to be cast as: `:= int32(v.(int))`?

tombuildsstuff commented on azurerm/resource_arm_app_service_plan.go:
```diff
@@ -0,0 +1,133 @@
+		return err
+	}
+	if read.ID == nil {
+		return fmt.Errorf("Cannot read Azure App Service Plan %s (resource group %s) ID", name, resGroup)
```
can we rename this `AzureRM`* for consistency?

tombuildsstuff commented on azurerm/resource_arm_app_service_plan.go:
```diff
@@ -0,0 +1,133 @@
+	}
+
+	d.Set("name", name)
+	d.Set("resource_group_name", resGroup)
```
can we set the other field values here?

tombuildsstuff commented on azurerm/resource_arm_app_service_plan_test.go:
```diff
@@ -0,0 +1,97 @@
+	}
+}
+
+var testAccAzureRMAppServicePlan_standard = `
```
for newer resources we're tending to make this a function which returns a formatted string:

```
func testAccAzureRMAppServicePlan_standard(rInt int) string {
  return fmt.Sprintf(`foo`, rInt)
}
```

tombuildsstuff commented on azurerm/resource_arm_app_service_plan_test.go:
```diff
@@ -0,0 +1,97 @@
+	"github.com/hashicorp/terraform/terraform"
+)
+
+func TestAccAzureRMAppServicePlan_standard(t *testing.T) {
```
Can we add another test case with all fields set?

tombuildsstuff commented on azurerm/resource_arm_app_service_plan.go:
```diff
@@ -0,0 +1,133 @@
+			"tier": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
```
Looking at the `SkuDescription` object:

```
// SkuDescription is description of a SKU for a scalable resource.
type SkuDescription struct {
	Name         *string       `json:"name,omitempty"`
	Tier         *string       `json:"tier,omitempty"`
	Size         *string       `json:"size,omitempty"`
	Family       *string       `json:"family,omitempty"`
	Capacity     *int32        `json:"capacity,omitempty"`
	SkuCapacity  *SkuCapacity  `json:"skuCapacity,omitempty"`
	Locations    *[]string     `json:"locations,omitempty"`
	Capabilities *[]Capability `json:"capabilities,omitempty"`
}
```

I think it makes sense to separate this out into it's own block i.e.:

```
sku {
  name = "Standard_S0"
  tier = "Standard"
  size = "S0"
  ...
}
```

Here's an [example of the Schema usage](https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/resource_arm_container_service.go#L48) - [expanding/parsing it out](https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/resource_arm_container_service.go#L474) and [flattening/writing it back](https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/resource_arm_container_service.go#L334) - which I'm hoping makes sense?

tombuildsstuff commented on azurerm/resource_arm_app_service_plan.go:
```diff
@@ -0,0 +1,133 @@
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceArmAppServicePlan() *schema.Resource {
```
could we add some documentation for this resource too?

lstolyarov commented:
Hey @tombuildsstuff, I am facing a problem when running the acceptance test for the App Service Plan. I get the following error: `Cannot find Web space acctestRG-2152939972729330496-WestEuropewebspace for subscription xxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx. ` Have you ever seen this problem before?

tombuildsstuff commented:
Hey @lstolyarov 

Taking a look at the request made in Debug mode, I can confirm I'm seeing the same response - which appears to be an Azure API / Service error:

```
PUT /subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/acctestRG-212/providers/Microsoft.Web/serverfarms/acctestAppServicePlan-test?api-version=2016-09-01 HTTP/1.1
Host: management.azure.com
User-Agent: HashiCorp-Terraform-v0.9.8
Content-Length: 91
Content-Type: application/json
Accept-Encoding: gzip

{
	"location": "westeurope",
	"properties": {},
	"sku": {
		"name": "S0",
		"tier": "Standard",
		"size": "S0"
	}
}
```

Which returns the response:
```
HTTP/1.1 404 Not Found
Content-Length: 667
...

{
	"Code": "NotFound",
	"Message": "Cannot find Web space acctestRG-212-WestEuropewebspace for subscription 00000000-0000-0000-0000-000000000000.",
	"Target": null,
	"Details": [{
		"Message": "Cannot find Web space acctestRG-212-WestEuropewebspace for subscription 00000000-0000-0000-0000-000000000000."
	}, {
		"Code": "NotFound"
	}, {
		"ErrorEntity": {
			"ExtendedCode": "03005",
			"MessageTemplate": "Cannot find Web space {0} for subscription {1}.",
			"Parameters": ["acctestRG-212-WestEuropewebspace", "00000000-0000-0000-0000-000000000000"],
			"Code": "NotFound",
			"Message": "Cannot find Web space acctestRG-212-WestEuropewebspace for subscription 00000000-0000-0000-0000-000000000000."
		}
	}],
	"Innererror": null
}
```

I've also tested in an HTTP client submitting via a POST rather than a PUT to check if this allowed is to be created - but unfortunately this doesn't appear to be an option. As such, I believe this is a service error, and the next step would be to [open an issue over on the Azure Rest API Specs Repository](https://github.com/Azure/azure-rest-api-specs/issues) which would allow the team to investigate further.

Hope that helps? :)

echuvyrov commented:
Hey @lstolyarov thanks for all your work on this. I'd love to help you resolve the issue so we can move your change along. I did a bit of looking and testing and it looks like you have "S0" as the plan for your tests. Using your code, I changed it to "S1" and the tests almost passed :) (got a different, hopefully familiar error). Here are a couple of links to take a look at:

App Service Plans: https://azure.microsoft.com/en-us/pricing/details/app-service/
Gist creating a resource group + web app plan using Go SDK (portions, including logging borrowed from TF): https://gist.github.com/echuvyrov/cd8f9c2fcbd90a01e7bace2c523bf915

Let me know if this helps or if there's something else going on here.

tombuildsstuff commented:
👋 hey @lstolyarov - have you had a chance to look into @echuvyrov's comments? :)

lstolyarov commented:
Hi @tombuildsstuff,

Apologies, I haven't had a chance yet.

@echuvyrov thanks for the suggestion. I am aiming to try it out this week.

Thanks

echuvyrov commented:
@lstolyarov let me know if you could use a helping hand - I think this is one of the key resources for Azure :)

lstolyarov commented:
Hey guys,

I just had a look at this and the app service plan does get created with the sku size set to "S1". However, the tests now fail when resources are being destroyed. The destroy command cannot execute because the resource still has a create operation running on it:

```
{
  "Code": "Conflict",
  "Message": "Cannot modify this web hosting plan because another operation is in progress. Details: Id: 096c1fa1-b248-4e67-98c5-f0e018b1b261, OperationName: Create, CreatedTime: 7/18/2017 5:42:06 PM, RequestId: 09e22889-8fb2-40f0-8446-2739e820f135, EntityType: 9",
  "Target": null,
  "Details": [
    {
      "Message": "Cannot modify this web hosting plan because another operation is in progress. Details: Id: 096c1fa1-b248-4e67-98c5-f0e018b1b261, OperationName: Create, CreatedTime: 7/18/2017 5:42:06 PM, RequestId: 09e22889-8fb2-40f0-8446-2739e820f135, EntityType: 9"
    },
    {
      "Code": "Conflict"
    },
    {
      "ErrorEntity": {
        "ExtendedCode": "59205",
        "MessageTemplate": "Cannot modify this web hosting plan because another operation is in progress. Details: {0}",
        "Parameters": [
          "Id: 096c1fa1-b248-4e67-98c5-f0e018b1b261, OperationName: Create, CreatedTime: 7/18/2017 5:42:06 PM, RequestId: 09e22889-8fb2-40f0-8446-2739e820f135, EntityType: 9"
        ],
        "Code": "Conflict",
        "Message": "Cannot modify this web hosting plan because another operation is in progress. Details: Id: 096c1fa1-b248-4e67-98c5-f0e018b1b261, OperationName: Create, CreatedTime: 7/18/2017 5:42:06 PM, RequestId: 09e22889-8fb2-40f0-8446-2739e820f135, EntityType: 9"
      }
    }
  ],
  "Innererror": null
}
```

@tombuildsstuff have you come across this before?

tombuildsstuff commented:
Hey @lstolyarov 

From what you've said above, I'd guess the SDK is returning from the `CreateOrUpdate` method when the resource creation has been submitted/requested, rather than polling for it's completion and returning once it's finished creating. In general I'd put that down to a Swagger bug, where the status code isn't right - let's ignore that for now, but we should _probably_ open a bug about it on the [azure-rest-api-specs](https://github.com/Azure/azure-rest-api-specs) repository at some point

To work around that - [in some resources we have functions which poll to check the resource has been created](https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/resource_arm_loadbalancer.go#L146) - which is what I'd suggest adding to the Create method.

Hopefully that should solve the issue you're seeing here :)

Thanks!

lstolyarov commented:
Hi @tombuildsstuff,

Thanks for the suggestion. I added the function but it doesn't help. The ProvisioningState is updated to "Succeeded" when the app service is created. I also tested this manually and I am only able to delete the app service plan about 5 minutes after it has been provisioned so there is definitely something else happening. I have got an issue open [here](https://github.com/Azure/azure-sdk-for-go/issues/667) updated with the latest problems.

echuvyrov commented:
hey @lstolyarov @tombuildsstuff I am running into the same issue you are. Looking into this into internally and hope to have an answer very soon.

tombuildsstuff commented:
Hey @echuvyrov 

Did you manage to get anywhere with this internally? :)

Thanks!

georgeedwards commented:
There's been an update on the SDK issue. Are we sure that `S0` is a valid tier?

marstr commented on azurerm/resource_arm_app_service_plan.go:
```diff
@@ -0,0 +1,216 @@
+		Sku: &sku,
+	}
+
+	_, error := AppServicePlanClient.CreateOrUpdate(resGroup, name, appServicePlan, make(chan struct{}))
```
You may as well update this to be:
``` Go
_, error := AppServicePlanClient.CreateOrUpdate(resGroup, name, appServicePlan, nil)
```
Instantiating a channel like this on the fly doesn't allow you a way to cancel and just adds a little garbage to be cleaned up later. :) Notice in [this playground example](https://play.golang.org/p/ycMN89XgqM) nil is passed to `wait(...)` but never triggers the cancel select case.

georgeedwards commented:
Hey @lstolyarov, Did you get a chance to look at the comments? Let me know if you need a hand - I would be really keen to see this as a supported resource.

echuvyrov commented:
hey @georgeedwards @lstolyarov @tombuildsstuff we should have this unblocked this week, will keep you posted. I will test at the end of the week or at the beginning of the next week to confirm.

echuvyrov commented:
hey @lstolyarov @tombuildsstuff @georgeedwards - our team reporting the acceptance tests passing now, can you please confirm that you are unblocked when you get a chance?

lstolyarov commented:
Hi @echuvyrov, thanks a lot for having a look at this. The tests are now passing:

```
$ TF_ACC=1 envchain azurecloudops go test ./azurerm -v -timeout 300m -parallel 5 -run TestAccAzureRMAppServicePlan
=== RUN   TestAccAzureRMAppServicePlan_importBasic
--- PASS: TestAccAzureRMAppServicePlan_importBasic (63.80s)
=== RUN   TestAccAzureRMAppServicePlan_importStandard
--- PASS: TestAccAzureRMAppServicePlan_importStandard (77.62s)
=== RUN   TestAccAzureRMAppServicePlan_importPremium
--- PASS: TestAccAzureRMAppServicePlan_importPremium (74.71s)
=== RUN   TestAccAzureRMAppServicePlan_basic
--- PASS: TestAccAzureRMAppServicePlan_basic (57.62s)
=== RUN   TestAccAzureRMAppServicePlan_standard
--- PASS: TestAccAzureRMAppServicePlan_standard (63.03s)
=== RUN   TestAccAzureRMAppServicePlan_premium
--- PASS: TestAccAzureRMAppServicePlan_premium (66.64s)
PASS
```
@tombuildsstuff I have added a couple more tests and updated the documentation. Is there anything else that needs to be done before this can be merged?

Thanks,
Leo

tombuildsstuff commented on azurerm/resource_arm_app_service_plan.go:
```diff
@@ -0,0 +1,215 @@
+				Set: resourceAzureRMAppServicePlanSkuHash,
+			},
+			"maximum_number_of_workers": {
+				Type:     schema.TypeString,
```
I think this should be an int rather than a String?

tombuildsstuff commented on azurerm/appserviceplan.go:
```diff
@@ -0,0 +1,18 @@
+
+		return res, string(res.AppServicePlanProperties.ProvisioningState), nil
+	}
+}
```
_minor_ could we move this into the `resource_arm_app_service_plan.go` file?

tombuildsstuff commented on azurerm/resource_arm_app_service_plan.go:
```diff
@@ -0,0 +1,215 @@
+		Timeout: 10 * time.Minute,
+	}
+	if _, err := stateConf.WaitForState(); err != nil {
+		return fmt.Errorf("Error waiting for App Service Plan (%s) to become available: %s", name, err)
```
_minor_ can we make this second formatting argument `%+v` rather than `%s` to show the full error?

tombuildsstuff commented on azurerm/appserviceplan.go:
```diff
@@ -0,0 +1,18 @@
+	return func() (interface{}, string, error) {
+		res, err := client.appServicePlansClient.Get(resourceGroupName, appserviceplan)
+		if err != nil {
+			return nil, "", fmt.Errorf("Error issuing read request in appServicePlanStateRefreshFunc to Azure ARM for App Service Plan '%s' (RG: '%s'): %s", appserviceplan, resourceGroupName, err)
```
can we make the final formatting argument `%+v` here to return the full error?

tombuildsstuff commented on azurerm/appserviceplan.go:
```diff
@@ -0,0 +1,18 @@
+	"github.com/hashicorp/terraform/helper/resource"
+)
+
+func appServicePlanStateRefreshFunc(client *ArmClient, resourceGroupName string, appserviceplan string) resource.StateRefreshFunc {
```
_really minor_ it'd be good to call `appserviceplan` -> `name` for consistency across these methods, if possible

tombuildsstuff commented on azurerm/resource_arm_app_service_plan.go:
```diff
@@ -0,0 +1,215 @@
+
+	properties := web.AppServicePlanProperties{}
+	if v, ok := d.GetOk("maximum_number_of_workers"); ok {
+		maximumNumberOfWorkers := v.(int32)
```
could we pull this out of the schema as an int, and then casting the result to an int32? i.e.

```
maximumNumberOfWorkers = int32(v.(int))
```

(whilst perhaps an odd-looking comment/ask - this should guarantee compatibility going forward)

tombuildsstuff commented on azurerm/resource_arm_app_service_plan.go:
```diff
@@ -0,0 +1,215 @@
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
```
since [this is in the documentation](https://github.com/terraform-providers/terraform-provider-azurerm/pull/1/files#diff-4fd01bbe4e96fffa5aff4ff7b99415f3R51) can we also add a field for tags? there's a helper schema for this i.e. `"tags": tagsSchema()"

tombuildsstuff commented on azurerm/resource_arm_app_service_plan.go:
```diff
@@ -0,0 +1,215 @@
+		Location:                 &location,
+		AppServicePlanProperties: &properties,
+		Sku: &sku,
+	}
```
(as above) can we add tags support here? it should be:
```
tags := d.Get("tags").(map[string]interface{})
expandedTags := expandTags(tags)
```

tombuildsstuff commented on azurerm/resource_arm_app_service_plan.go:
```diff
@@ -0,0 +1,215 @@
+		F: resourceAzureRMAppServicePlanSkuHash,
+	}
+
+	sku := make(map[string]interface{}, 3)
```
_minor_ I think this can become `2` rather than `3`?

tombuildsstuff commented on azurerm/resource_arm_app_service_plan_test.go:
```diff
@@ -0,0 +1,179 @@
+	return fmt.Sprintf(`
+resource "azurerm_resource_group" "test" {
+    name = "acctestRG-%d"
+    location = "West Europe"
```
can we swap the value to `"%s"`?

tombuildsstuff commented on azurerm/resource_arm_app_service_plan_test.go:
```diff
@@ -0,0 +1,179 @@
+			size = "B1"
+		}
+}
+`, rInt, rInt)
```
can we add `location` as the second argument here?

tombuildsstuff commented on website/azurerm.erb:
```diff
@@ -42,6 +42,17 @@
             </li>
 
+            <li<%= sidebar_current("docs-azurerm-resource-app-service") %>>
+              <a href="#">App Service Resources</a>
```
🤔 do you think it's worth making this "App Service (Web Apps) Resources", like we've done for CosmosDB/DocumentDB?

lstolyarov replied:
Good point. Yes, I think it is worth doing that.

tombuildsstuff commented on website/azurerm.erb:
```diff
@@ -42,6 +42,17 @@
+              <ul class="nav nav-visible">
+
+                <li<%= sidebar_current("docs-azurerm-resource-app-service-plan") %>>
+                  <a href="/docs/providers/azurerm/r/app_service_pla.html">azurerm_app_service_plan</a>
```
`app_service_pla.html` -> `app_service_plan.html`

tombuildsstuff commented on website/docs/r/app_service_plan.html.markdown:
```diff
@@ -0,0 +1,65 @@
+* `tier` - (Required) Specified the plan's pricing tier.
+* `size` - (Required) Specified the plan's instance size. 
+
+* `maximum_number_of_workers` - (Optional) .
```
could we add a description for this field, here's the description from the Azure docs: "Maximum number of instances that can be assigned to this App Service plan."

tombuildsstuff commented on azurerm/resource_arm_app_service_plan_test.go:
```diff
@@ -0,0 +1,179 @@
+
+resource "azurerm_app_service_plan" "test" {
+    name                = "acctestASP-%d"
+    location            = "West Europe"
```
can we make this `${azurerm_resource_group.test.location}` to only insert the location formatting argument a single time?

tombuildsstuff commented:
Tests pass after the rebase:

```
$ acctests azurerm TestAccAzureRMAppServicePlan_
=== RUN   TestAccAzureRMAppServicePlan_importBasic
--- PASS: TestAccAzureRMAppServicePlan_importBasic (78.84s)
=== RUN   TestAccAzureRMAppServicePlan_importStandard
--- PASS: TestAccAzureRMAppServicePlan_importStandard (80.06s)
=== RUN   TestAccAzureRMAppServicePlan_importPremium
--- PASS: TestAccAzureRMAppServicePlan_importPremium (80.61s)
=== RUN   TestAccAzureRMAppServicePlan_importComplete
--- PASS: TestAccAzureRMAppServicePlan_importComplete (79.25s)
=== RUN   TestAccAzureRMAppServicePlan_basic
--- PASS: TestAccAzureRMAppServicePlan_basic (77.02s)
=== RUN   TestAccAzureRMAppServicePlan_standard
--- PASS: TestAccAzureRMAppServicePlan_standard (74.75s)
=== RUN   TestAccAzureRMAppServicePlan_premium
--- PASS: TestAccAzureRMAppServicePlan_premium (77.72s)
=== RUN   TestAccAzureRMAppServicePlan_complete
--- PASS: TestAccAzureRMAppServicePlan_complete (74.66s)
PASS
ok  	github.com/terraform-providers/terraform-provider-azurerm/azurerm	622.930s
```

tombuildsstuff commented:
Updated to make the number of workers configurable - tests pass:

```
$ acctests azurerm TestAccAzureRMAppServicePlan_

=== RUN   TestAccAzureRMAppServicePlan_importBasic
--- PASS: TestAccAzureRMAppServicePlan_importBasic (69.20s)
=== RUN   TestAccAzureRMAppServicePlan_importStandard
--- PASS: TestAccAzureRMAppServicePlan_importStandard (74.10s)
=== RUN   TestAccAzureRMAppServicePlan_importPremium
--- PASS: TestAccAzureRMAppServicePlan_importPremium (61.45s)
=== RUN   TestAccAzureRMAppServicePlan_importComplete
--- PASS: TestAccAzureRMAppServicePlan_importComplete (77.09s)
=== RUN   TestAccAzureRMAppServicePlan_basic
--- PASS: TestAccAzureRMAppServicePlan_basic (60.22s)
=== RUN   TestAccAzureRMAppServicePlan_standard
--- PASS: TestAccAzureRMAppServicePlan_standard (79.58s)
=== RUN   TestAccAzureRMAppServicePlan_premium
--- PASS: TestAccAzureRMAppServicePlan_premium (76.44s)
=== RUN   TestAccAzureRMAppServicePlan_premiumUpdated
--- PASS: TestAccAzureRMAppServicePlan_premiumUpdated (90.97s)
=== RUN   TestAccAzureRMAppServicePlan_complete
--- PASS: TestAccAzureRMAppServicePlan_complete (76.79s)
PASS
ok  	github.com/terraform-providers/terraform-provider-azurerm/azurerm	665.873s
```

georgeedwards commented:
@lstolyarov Are ther any plans to support adding env variables to the configuration?
